onnxruntime/core/optimizer/constant_sharing.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef ENABLE_TRAINING
+
+#include <limits>
+
+#include "core/framework/op_kernel.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/constant_sharing.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+
+/**
+ * @brief Share initializer for those who hold same value in same type and shape.
+ *
+ * @tparam T The initializer value type, used to retrieved data values.
+ * @param graph Target graph to iterate.
+ * @param node Target node to check initializer input.
+ * @param input_index Input index of target node.
+ * @param type_value_plus_rank_to_shared_arg_map Accumulated map from value/rank to initializer.
+ *  The key indiciate initializer's data type, value and rank.
+ *  The value is the first initializer NodeArg* to be shared.
+ * @param data_type Initializer data type.
+ * @return true If the input initializer can be shared.
+ * @return false If the input initializer CANNOT be shared.
+ */
+template <typename T>
+bool ConstantSharing::ShareInitializer(Graph& graph, Node* node, int input_index,
+                                       std::map<std::string, NodeArg*>&
+                                           type_value_plus_rank_to_shared_arg_map,
+                                       int32_t data_type) const {
+  InlinedVector<T> values;
+  const bool require_constant = true;
+  const NodeArg* input_def = node->InputDefs()[input_index];
+  if (!optimizer_utils::AppendTensorFromInitializer(graph, *input_def, values, require_constant) ||
+      values.size() != 1) {
```

Comment:
What's the reason we append to values? Doesn't that do unnecessary copies (potentially of a large initializer) vs. using Initializer directly.

Naively I would have expected it to be something like the below so the cheapest/easiest checks are first:

- check input def Shape has size of 1 (rank could be 0 or 1)
- check input def type is handled
- if so, create an Initializer instance to read the value

Instead of re-using the first initializer found with a given value and rank, it may be clearer to create a new one with the value you create in `p`. that way someone debugging doesn't see a misleading name for the value at the replacement sites.